firmware/src/vcp.rs

```diff
@@ -0,0 +1,231 @@
+// Copyright 2019-2022 Alexis Marquet   
+// Dual licensed under the Apache 2.0 and MIT licenses.
+
+use core::cmp::Ordering;
+
+
+use crate::bsp::{dma::DMA, gpio::{Pins}, rcc::Clocks};
+
+use stm32ral::usart;
+use stm32ral::{modify_reg, write_reg};
+use usbd_serial::{ParityType, StopBits};
+
+/// UART configuration struct
+#[derive(PartialEq, Eq, Clone, Copy)]
+pub struct VcpConfig {
+    pub stop_bits: StopBits,
+    pub data_bits: u8,
+    pub parity_type: ParityType,
+    pub data_rate: u32,
+}
+
+impl Default for VcpConfig {
+    fn default() -> Self {
+        VcpConfig {
+            stop_bits: StopBits::One,
+            data_bits: 8,
+            parity_type: ParityType::None,
+            data_rate: 8_000,
+        }
+    }
+}
+
+pub struct VCP<'a> {
+    uart: usart::Instance,
+    pins: &'a Pins<'a>,
+    dma: &'a DMA,
+    rx_buffer: [u8; 256],
+    tx_buffer: [u8; 256],
```

Comment:
USB packet size is 512 bytes, but this buffer is just 256 bytes: the firmware will panic after receiving a large packet.